### ISSUE TITLE
Linewise % motion not respected when matchit.vim plugin sourced.

### ISSUE THREAD

**Issue by hardenedapple** [neovim/neovim#5691]
- `nvim --version`:
NVIM 0.1.6
- Vim (version: ) behaves differently?
Vim 8.0  -- Yes behaves as expected
- Operating system/version:
Arch Linux
- Terminal name/version:
XTerm(327)
- `$TERM`:
xterm-256color

### Actual behaviour
With the test file
```
function_call(argument1, argument2,
                argument3)
```

Start with `nvim -u NORC test.txt`

Run the commands
```
/call<CR>
yV%jp
```

Text in buffer is now
```
function_call(argument1, argument2,
          call(argument1, argument2,
                argument3)      argument3)
```
### Expected behaviour

Following those commands, the text in the buffer should be
```
function_call(argument1, argument2,
                argument3)
function_call(argument1, argument2,
                argument3)
```

This is what Vim does

### See that the problem is in matchit.vim
If start nvim with `nvim -u NONE test.txt` and run the commands above, get expected results (i.e. same as vim).

In comparison, under `nvim -u NONE test.txt`.
```
source /usr/share/nvim/runtime/plugin/matchit.vim
/call<CR>
yV%jp
```

demonstrate the problem.

### General other information
Similarly, the by-default linewise operation on `#if/#else` blocks that vim does is not respected once the `matchit.vim` plugin is sourced.

**Comment by justinmk** [comment]
So, a bug in matchit.vim ?

**Comment by jszakmeister** [comment]
It does look like there have been some updates to matchit.vim.  Here's a diff between the Vim master and Neovim master:

```diff
diff --git i/runtime/plugin/matchit.vim w/runtime/plugin/matchit.vim
index c0f1f08..22acb06 100644
--- i/runtime/plugin/matchit.vim
+++ w/runtime/plugin/matchit.vim
@@ -1,8 +1,9 @@
 "  matchit.vim: (global plugin) Extended "%" matching
-"  Last Change: Fri Jul 29 01:20 AM 2016 EST
+"  Last Change: 2016 Aug 21
 "  Maintainer:  Benji Fisher PhD   <benji@member.AMS.org>
 "  Version:     1.13.2, for Vim 6.3+
-"		Fix from Tommy Allen included.
+"		Fix from Fernando Torres included.
+"		Improvement from Ken Takata included.
 "  URL:		http://www.vim.org/script.php?script_id=39
 
 " Documentation:
@@ -44,6 +45,7 @@ endif
 let loaded_matchit = 1
 let s:last_mps = ""
 let s:last_words = ":"
+let s:patBR = ""
 
 let s:save_cpo = &cpo
 set cpo&vim
@@ -121,8 +123,8 @@ function! s:Match_wrapper(word, forward, mode) range
     execute "let match_words =" b:match_words
   endif
 " Thanks to Preben "Peppe" Guldberg and Bram Moolenaar for this suggestion!
-  if (match_words != s:last_words) || (&mps != s:last_mps) ||
-    \ exists("b:match_debug")
+  if (match_words != s:last_words) || (&mps != s:last_mps)
+      \ || exists("b:match_debug")
     let s:last_mps = &mps
     " The next several lines were here before
     " BF started messing with this script.
@@ -148,6 +150,10 @@ function! s:Match_wrapper(word, forward, mode) range
     if exists("b:match_debug")
       let b:match_pat = s:pat
     endif
+    " Reconstruct the version with unresolved backrefs.
+    let s:patBR = substitute(match_words.',',
+      \ s:notslash.'\zs[,:]*,[,:]*', ',', 'g')
+    let s:patBR = substitute(s:patBR, s:notslash.'\zs:\{2,}', ':', 'g')
   endif
 
   " Second step:  set the following local variables:
@@ -192,14 +198,10 @@ function! s:Match_wrapper(word, forward, mode) range
   " group = colon-separated list of patterns, one of which matches
   "       = ini:mid:fin or ini:fin
   "
-  " Reconstruct the version with unresolved backrefs.
-  let patBR = substitute(match_words.',',
-    \ s:notslash.'\zs[,:]*,[,:]*', ',', 'g')
-  let patBR = substitute(patBR, s:notslash.'\zs:\{2,}', ':', 'g')
   " Now, set group and groupBR to the matching group: 'if:endif' or
   " 'while:endwhile' or whatever.  A bit of a kluge:  s:Choose() returns
   " group . "," . groupBR, and we pick it apart.
-  let group = s:Choose(s:pat, matchline, ",", ":", prefix, suffix, patBR)
+  let group = s:Choose(s:pat, matchline, ",", ":", prefix, suffix, s:patBR)
   let i = matchend(group, s:notslash . ",")
   let groupBR = strpart(group, i)
   let group = strpart(group, 0, i-1)
@@ -255,7 +257,12 @@ function! s:Match_wrapper(word, forward, mode) range
 
   " Fifth step:  actually start moving the cursor and call searchpair().
   " Later, :execute restore_cursor to get to the original screen.
-  let view = winsaveview()
+  let restore_cursor = virtcol(".") . "|"
+  normal! g0
+  let restore_cursor = line(".") . "G" .  virtcol(".") . "|zs" . restore_cursor
+  normal! H
+  let restore_cursor = "normal!" . line(".") . "Gzt" . restore_cursor
+  execute restore_cursor
   call cursor(0, curcol + 1)
   " normal! 0
   " if curcol
@@ -269,7 +276,7 @@ function! s:Match_wrapper(word, forward, mode) range
   let sp_return = searchpair(ini, mid, fin, flag, skip)
   let final_position = "call cursor(" . line(".") . "," . col(".") . ")"
   " Restore cursor position and original screen.
-  call winrestview(view)
+  execute restore_cursor
   normal! m'
   if sp_return > 0
     execute final_position
@@ -630,7 +637,7 @@ endfun
 " idea to give it its own matching patterns.
 fun! s:MultiMatch(spflag, mode)
   if !exists("b:match_words") || b:match_words == ""
-    return {}
+    return ""
   end
   let restore_options = (&ic ? "" : "no") . "ignorecase"
   if exists("b:match_ignorecase")
@@ -656,6 +663,7 @@ fun! s:MultiMatch(spflag, mode)
     \ exists("b:match_debug")
     let s:last_words = match_words
     let s:last_mps = &mps
+    let match_words = match_words . (strlen(match_words) ? "," : "") . default
     if match_words !~ s:notslash . '\\\d'
       let s:do_BR = 0
       let s:pat = match_words
@@ -663,8 +671,8 @@ fun! s:MultiMatch(spflag, mode)
       let s:do_BR = 1
       let s:pat = s:ParseWords(match_words)
     endif
-    let s:all = '\%(' . substitute(s:pat . (strlen(s:pat)?",":"") . default,
-      \	'[,:]\+','\\|','g') . '\)'
+    let s:all = '\%(' . substitute(s:pat . (strlen(s:pat) ? "," : "") . default,
+	\ '[,:]\+', '\\|', 'g') . '\)'
     if exists("b:match_debug")
       let b:match_pat = s:pat
     endif
@@ -690,7 +698,15 @@ fun! s:MultiMatch(spflag, mode)
     let skip = 's:comment\|string'
   endif
   let skip = s:ParseSkip(skip)
-  let view = winsaveview()
+  " let restore_cursor = line(".") . "G" . virtcol(".") . "|"
+  " normal! H
+  " let restore_cursor = "normal!" . line(".") . "Gzt" . restore_cursor
+  let restore_cursor = virtcol(".") . "|"
+  normal! g0
+  let restore_cursor = line(".") . "G" .  virtcol(".") . "|zs" . restore_cursor
+  normal! H
+  let restore_cursor = "normal!" . line(".") . "Gzt" . restore_cursor
+  execute restore_cursor
 
   " Third step: call searchpair().
   " Replace '\('--but not '\\('--with '\%(' and ',' with '\|'.
@@ -708,14 +724,14 @@ fun! s:MultiMatch(spflag, mode)
   while level
     if searchpair(openpat, '', closepat, a:spflag, skip) < 1
       call s:CleanUp(restore_options, a:mode, startline, startcol)
-      return {}
+      return ""
     endif
     let level = level - 1
   endwhile
 
-  " Restore options and return view dict to restore the original position.
+  " Restore options and return a string to restore the original position.
   call s:CleanUp(restore_options, a:mode, startline, startcol)
-  return view
+  return restore_cursor
 endfun
 
 " Search backwards for "if" or "while" or "<tag>" or ...
```

**Comment by jszakmeister** [comment]
I'd put up a PR, but I've really got to run.

**Comment by chrisbra** [comment]
I would keep the winsaveview() stuff. The current jumping is nowadays 
just ugly.

It seems, this has been included in neovim but is still waiting to be 
included in vim: vim/vim#955

Best,
Christian
-- 
Wer sich nicht einmal am Tag irrt, ist entweder dumm oder phantasielos oder
hat vom Tag nichts gehabt.

**Comment by chrisbra** [comment]
@hardenedapple How did you test on Vim and did you enable matchit there? Because I see the same behaviour in Vim. So this seems like a general bug in the matchit plugin. And the reason is obvious:

```viml
onoremap <silent> %  v:<C-U>call <SID>Match_wrapper('',1,'o') <CR>
onoremap <silent> g% v:<C-U>call <SID>Match_wrapper('',0,'o') <CR>
```
(e.g. it hard coded uses 'v' in operator pending mode)

**Comment by hardenedapple** [comment]
@chrisbra

Apologies, that was a bad description on my part.
I didn't test Vim with the matchit plugin at all - I was talking about the Vim behaviour just to show what should happen (comparing the two behaviours without personal configuration).

 After re-reading my post and the following comments I see how I was confusing.

**Comment by mhinz** [comment]
@chrisbra This has been fixed in Vim, right? (We both use matchit 1.17.)

**Comment by chrisbra** [comment]
Yed
